cprogs/acctdump.8
.TH acctdump 8 2014-10-06 chiark-utils-bin
.SH NAME
acctdump \- accounting data dump utility
.SH SYNOPSIS
.B acctdump
.RB [ -f|--forwards ]
.RB [ -q|--no-banner ]
.RB [ -p|--stdin ]
.RB [ -r|--raw ]
.RB [ -u|--resource ]
.RI [ <file> ... ]
.SH DESCRIPTION
.B acctdump
reads process accounting data from 
.B file
or, by default, from the system process account file 
.I /var/log/account/pacct 
and outputs it as a table on stdout.
.SH OPTIONS
.TP
.B -f|--forwards
Output data with the oldest entry first rather than last.
.TP
.B -q|--no-banner
Don\'t output a header in the first line.
.TP
.B -p|--stdin
Read accounting data from stdin.
.TP
.B -r|--raw
Don\'t convert date, uid, gid, tty, and exit status to human readable forms.
.TP
.B -u|--resource
Additionally display resource usage data.
.SH "SEE ALSO"
.BR ac (1)
.BR accton (8)
.BR dump-acct (8)
.BR dump-utmp (8)
.BR lastcomm (1)
.BR sa (8)
.SH AUTHOR
.B acctdump
is
.BR
Copyright (C) 1998, 2001, 2006 Ian Jackson <ian@chiark.greenend.org.uk>

This manpage was written by Jonathan Amery.  It is
.BR
Copyright (C) 2014 Jonathan Amery <jdamery@chiark.greenend.org.uk>

This is free software, distributed under the GNU General Public
Licence, version 3 or (at your option) any later version; see
/usr/share/doc/chiark-utils-bin/copyright or
/usr/share/common-licenses/GPL-3
for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/acctdump.c
/*
 * acctdump.c - accounting data dump utility
 *
 * Copyright (C) 1998 Ian Jackson <ian@chiark.greenend.org.uk>
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this file; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdarg.h>
#include <wait.h>
#include <string.h>
#include <pwd.h>
#include <grp.h>
#include <dirent.h>
#include <ctype.h>
#include <errno.h>
#include <sys/stat.h>

typedef unsigned long long u64;


/* Sadly this thing is not very portable */

#if defined(__linux__)

#include <sys/types.h>
#include <sys/acct.h>

typedef struct acct_v3 struct_acct;
#define HAVE_AC_EXITCODE
#define HAVE_AC_FLT
#define FIELD_AC_FLAG(as) ((as)->ac_flag)

#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)

#include <sys/param.h>
#include <sys/types.h>
#include <sys/acct.h>

typedef struct acctv2 struct_acct;
#define HAVE_AC_IO_MEM
#define FIELD_AC_FLAG(as) ((as)->ac_flagx & ~ANVER)

#else

#error Do not know what struct_acct to use on this platform

#endif


#include "myopt.h"

static int forwards, nobanner, usestdin, raw, usages;

static int de_used, de_allocd;
static struct deventry {
  const char *fn;
  dev_t dev;
} *deventries;

static const struct cmdinfo cmdinfos[]= {
  { "--forwards",  'f',  0, &forwards, 0, 0, 1, 0, 0 },
  { "--no-banner", 'q',  0, &nobanner, 0, 0, 1, 0, 0 },
  { "--stdin",     'p',  0, &usestdin, 0, 0, 1, 0, 0 },
  { "--raw",       'r',  0, &raw,      0, 0, 1, 0, 0 },
  { "--resource",  'u',  0, &usages,   0, 0, 1, 0, 0 },
  {  0                                                 }
};

#ifdef HAVE_AC_EXITCODE
static const char *sigabbrev[]= {
  "HUP", "INT", "QUIT", "ILL", "TRAP", "ABRT", "BUS", "FPE",
  "KILL", "USR1", "SEGV", "USR2", "PIPE", "ALRM", "TERM", "STKFLT",
  "CHLD", "CONT", "STOP", "TSTP", "TTIN", "TTOU", "URG", "XCPU",
  "XFSZ", "VTALRM", "PROF", "WINCH", "IO"
};
#endif

void usagemessage(void) {
  fputs("usage: acctdump [<options>] [<file> ...]\n"
	"options: -f|--forwards -q|--no-banner -p|--stdin -r|--raw -u|--resource\n",
	stderr);
  if (ferror(stderr)) { perror("print usage"); exit(8); }
}

static void checkstdout(void) {
  if (ferror(stdout)) { perror("stdout"); exit(8); }
}

static void scandev(const char *basename, int levelsleft) {
  /* We deliberately ignore most errors */
  DIR *dir;
  struct dirent *de;
  struct stat stab;
  int fnbufalloc, fnbufreq, r, basel, nallocd;
  char *fnbuf, *nfnbuf;
  struct deventry *ndeventries;
  
  if (levelsleft==0) return;

  dir= opendir(basename); 
  if (!dir) {
    fprintf(stderr, "%s: opendir: %s\n", basename, strerror(errno));
    return;
  }
  fnbufalloc= 0;
  fnbuf= 0;
  basel= strlen(basename);

  while ((errno=0, de= readdir(dir))) {
    fnbufreq= basel+1+strlen(de->d_name)+1;
    if (fnbufalloc<fnbufreq) {
      fnbufalloc= fnbufreq+10;
      nfnbuf= realloc(fnbuf,fnbufalloc);
      if (!nfnbuf) { free(fnbuf); fnbufalloc=0; continue; }
      fnbuf= nfnbuf;
    }
    sprintf(fnbuf,"%s/%s",basename,de->d_name);
    r= lstat(fnbuf,&stab);
    if (r) {
      fprintf(stderr, "%s: %s\n", fnbuf, strerror(errno));
      continue;
    }
    if (S_ISCHR(stab.st_mode)) {
      if (de_used >= de_allocd) {
	nallocd= (de_allocd+10)<<1;
	ndeventries= realloc(deventries,nallocd*sizeof(*deventries));
	if (!ndeventries) continue;
	de_allocd= nallocd;
	deventries= ndeventries;
      }
      deventries[de_used].fn= strdup(fnbuf+5); /* remove /dev */
      if (!deventries[de_used].fn) continue;
      deventries[de_used].dev= stab.st_rdev;
      de_used++;
    } else if (S_ISDIR(stab.st_mode) && de->d_name[0] != '.') {
      scandev(fnbuf,levelsleft-1);
    }
  }
  if (errno)
      fprintf(stderr, "%s: readdir: %s\n", basename, strerror(errno));
  closedir(dir);
  free(fnbuf);
}

static int walkdev_cptr(const void *av, const void *bv) {
  const struct deventry *a= av;
  const struct deventry *b= bv;
  return a->dev - b->dev;
}
  
static void printbanner(void) {
  if (raw) {
    fputs("begin date command          "
	  "uid      gid      tty dev  FSDX "
#ifdef HAVE_AC_EXITCODE
	  "exit"
#endif
	  , stdout);
  } else {
    fputs("begin date and time command          "
	  "user     group    tty dev    FSDX "
#ifdef HAVE_AC_EXITCODE
	  "sigexit"
#endif
	  , stdout);
  }
  if (usages) {
    fputs("  user time   sys time  elap time "
#ifdef HAVE_AC_FLT
	  "  minflt   maxflt"
#endif
#ifdef HAVE_AC_IO_MEM
	  "  avg.mem      io"
#endif
	  , stdout);
  }
  putchar('\n');
  checkstdout();
}

static void printrecord(const struct_acct *as, const char *filename) {
  static int walkeddev;

  int i, r;
  const char *fp;
  char buf[100];
  struct tm *tm;
  struct deventry *deve, devlookfor;
  struct passwd *pw;
  struct group *gr;
  time_t btime;
  char commbuf[sizeof(as->ac_comm)];

  if (raw) {
    printf("%10lu ",(unsigned long)as->ac_btime);
  } else {
    btime= as->ac_btime;
    tm= localtime(&btime);
    if (tm) {
      strftime(buf,sizeof(buf),"%Y-%m-%d %H:%M:%S",tm); buf[sizeof(buf)-1]= 0;
    } else {
      snprintf(buf,sizeof(buf),"@%lu",(unsigned long)btime);
    }
    printf("%19s ",buf);
  }

  for (i=0; i<sizeof(as->ac_comm); i++) {
    int c=as->ac_comm[i];
    commbuf[i]= ((c!=0 && c<=32) || c>=127) ? '?' : c;
  }
  printf("%-*.*s ", (int)sizeof(commbuf),(int)sizeof(commbuf), commbuf);
  
  pw= raw ? 0 : getpwuid(as->ac_uid);
  if (pw) printf("%-8s ",pw->pw_name);
  else printf("%-8ld ",(long)as->ac_uid);
  
  gr= raw ? 0 : getgrgid(as->ac_gid);
  if (gr) printf("%-8s ",gr->gr_name);
  else printf("%-8ld ",(long)as->ac_gid);

  if (raw) {
    if (!(as->ac_tty + 1) /* check for -1 without knowing type */) {
      printf("-        ");
    } else {
      printf("%08lx ",(unsigned long)as->ac_tty);
    }
  } else {
    if (!(as->ac_tty + 1)) {
      printf("-          ");
    } else {
      if (!walkeddev) {
	scandev("/dev",4);
	qsort(deventries,de_used,sizeof(*deventries),walkdev_cptr);
	walkeddev= 1;
      }
      devlookfor.fn= 0;
      devlookfor.dev= as->ac_tty;
      deve= bsearch(&devlookfor,deventries,de_used,sizeof(*deventries),walkdev_cptr);
      if (deve) {
	printf("%-10s ",deve->fn);
      } else {
	printf("%08lx   ",(unsigned long)as->ac_tty);
      }
    }
  }

  r= FIELD_AC_FLAG(as);
  for (i=1, fp= "FS4DX"; *fp; fp++, i<<=1) {
    if (r&i) {
      putchar(*fp);
      r &= ~i;
    } else if (!isdigit(*fp)) {
      putchar(' ');
    }
  }
  if (r) {
    printf("#%x",r);
  }
  putchar(' ');

#ifdef HAVE_AC_EXITCODE
  int dc;
  dc= WCOREDUMP(as->ac_exitcode) ? 'd' : 'k';
  if (raw) {
    if (WIFEXITED(as->ac_exitcode)) {
      printf(" %3d",WEXITSTATUS(as->ac_exitcode));
    } else if (WIFSIGNALED(as->ac_exitcode)) {
      printf("%c%3d",
	     dc,
	     WTERMSIG(as->ac_exitcode));
    } else {
      printf("%04lx",(unsigned long)as->ac_exitcode);
    }
  } else {
    if (WIFEXITED(as->ac_exitcode)) {
      printf(" %6d",WEXITSTATUS(as->ac_exitcode));
    } else if (WIFSIGNALED(as->ac_exitcode)) {
      r= WTERMSIG(as->ac_exitcode);
      if (r>0 && r<=sizeof(sigabbrev)/sizeof(*sigabbrev)) {
	printf("%c%6s",
	       dc,
	       sigabbrev[r-1]);
      } else {
	printf("%cSIG%-3d",
	       dc,
	       r);
      }
    } else {
      printf("#%04lx",(unsigned long)as->ac_exitcode);
    }
  }
#endif /*HAVE_AC_EXITCODE*/

  if (usages) {
    printf(" %10lu %10lu %10lu",
	   (unsigned long)as->ac_utime,
	   (unsigned long)as->ac_stime,
	   (unsigned long)as->ac_etime);
#ifdef HAVE_AC_FLT
    printf(" %8lu %8lu",
	   (unsigned long)as->ac_minflt,
	   (unsigned long)as->ac_majflt);
#endif
#ifdef HAVE_AC_IO_MEM
    printf(" %4e %4e",
	   as->ac_mem,
	   as->ac_io);
#endif
  }
  putchar('\n');

  checkstdout();
}

static void processfile(FILE *file, const char *filename) {
  struct_acct as;
  long pos;
  int r;
  
  if (forwards) {
    while ((r= fread(&as,1,sizeof(as),file)) == sizeof(as)) {
      printrecord(&as,filename);
    }
  } else {
    r= fseek(file,0,SEEK_END); if (r) { perror(filename); exit(8); }
    pos= ftell(file); if (pos==-1) { perror(filename); exit(8); }
    if (pos % sizeof(as)) { 
      fprintf(stderr, "%s: File size is not an integral number "
	      "of accounting records\n", filename);
      exit(8);
    }
    for (;;) {
      if (pos<sizeof(as)) break;
      pos -= sizeof(as);
      r= fseek(file,pos,SEEK_SET); if (r==-1) { perror(filename); exit(8); }
      r= fread(&as,1,sizeof(as),file); if (r!=sizeof(as)) { perror(filename); exit(8); }
      printrecord(&as,filename);
    }
  }
  if (ferror(file) || fclose(file)) { perror(filename); exit(8); }
}

static void processnamedfile(const char *filename) {
  FILE *file;

  file= fopen(filename,"rb"); if (!file) { perror(filename); exit(8); }
  processfile(file,filename);
}

int main(int argc, const char *const *argv) {
  myopt(&argv,cmdinfos);
  if (!nobanner) printbanner();
  if (usestdin) {
    processfile(stdin,"<standard input>");
  } else if (!*argv) {
    processnamedfile("/var/log/account/pacct");
  } else {
    while (*argv) {
      processnamedfile(*argv);
      argv++;
    }
  }
  checkstdout();
  if (fflush(stdout)) { perror("flush stdout"); exit(8); }
  return 0;
}
